internal/config/storage.go

```diff
@@ -59,6 +64,64 @@ type Local struct {
 
 // Git contains configuration for referencing a git repository.
 type Git struct {
-	Repository string `json:"repository,omitempty" mapstructure:"repository"`
-	Ref        string `json:"ref,omitempty" mapstructure:"ref"`
+	Repository     string         `json:"repository,omitempty" mapstructure:"repository"`
+	Ref            string         `json:"ref,omitempty" mapstructure:"ref"`
+	Authentication Authentication `json:"authentication,omitempty" mapstructure:"authentication"`
+}
+
+// Authentication holds structures for various types of auth we support.
+// Token auth will take priority over Basic auth if both are provided.
+//
+// To make things easier, if there are multiple inputs that a particular auth method needs, and
+// not all inputs are given but only partially, we will return a validation error.
+// (e.g. if username for basic auth is given, and token is also given a validation error will be returned)
+type Authentication struct {
+	Enabled   bool       `json:"enabled,omitempty" mapstructure:"enabled"`
+	BasicAuth *BasicAuth `json:"basic,omitempty" mapstructure:"basic"`
+	TokenAuth *TokenAuth `json:"token,omitempty" mapstructure:"token"`
+}
+
+func (a *Authentication) validate() error {
+	if a.Enabled {
+		if a.BasicAuth == nil && a.TokenAuth == nil {
+			return errors.New("authentication enabled but no auth inputs provided")
+		}
+
+		if a.BasicAuth != nil {
+			if err := a.BasicAuth.validate(); err != nil {
+				return err
+			}
+		}
+
+		if a.TokenAuth != nil {
+			if err := a.TokenAuth.validate(); err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
If we drop enabled, we could just do something like:
```suggestion
	if a.BasicAuth != nil && a.TokenAuth != nil {
		return errors.New("git authentication: cannot enable both basic and token")
	}

	if a.BasicAuth != nil {
		if err := a.BasicAuth.validate(); err != nil {
			return err
		}
	}

	if a.TokenAuth != nil {
		if err := a.TokenAuth.validate(); err != nil {
			return err
		}
	}
```